### PR TITLE
Refactor to enable support for non-form based multipart requests

### PR DESCRIPTION
**Pull request by adamretter**
As discussed here - http://mail-archives.apache.org/mod_mbox/hc-httpclient-users/201906.mbox/%3Cm21s0dfas4.fsf%40nwalsh.com%3E

### CODE REVIEW

**Comment by ok2c** [comment]
@adamretter 
1. There are failing test cases
2. Can `MultipartPart` be condensed to `MultiPart`. Multi-part-part sounds funny to me but I am not a native English speaker, so feel free to disregard.

I'll do a more thorough review tomorrow.

File: httpclient5/src/main/java/org/apache/hc/client5/http/entity/mime/AbstractMultipartPartBuilder.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+ */
+
+package org.apache.hc.client5.http.entity.mime;
+
+/**
+ * Abstract Builder for individual {@link org.apache.hc.client5.http.entity.mime.MultipartPart}s.
+ *
+ * @since 5.0
+ */
+public abstract class AbstractMultipartPartBuilder<T extends MultipartPart, S extends AbstractMultipartPartBuilder<T, S>> {
```
Comment (ok2c) [inline httpclient5/src/main/java/org/apache/hc/client5/http/entity/mime/AbstractMultipartPartBuilder.java]: @adamretter I would not create an abstract class just to re-use two instance variables. Would you mind getting rid of it? 

All other things look good to me. If you could fix the failing test cases and I would merge the change-set. 

Some test coverage for `MultipartPartBuilder` would also be nice.

**Comment by adamretter** [comment]
@ok2c Okay I have made the changes you asked for and added some tests

**Comment by ok2c** [comment]
@adamretter Committed as d71d0f5a4cecf0c620f9110cb567524416100895 with some minor tweaks.

**Pull request closed without merging**